### PR TITLE
Breaking: Create eslint:recommended and add to --init (fixes #2713)

### LINKED ISSUE #2713: Create "eslint:recommended" to extend from
As part of the 1.0.0 change to use `--reset`, we need to define "eslint:recommended" as a way to back in recommended defaults. 

- [ ] update eslint.json so only "potential errors" are on by default
- [ ] add "eslint:recommended" as a hardcoded option for `extends` that merges in eslint.json
- [ ] update `--init` to automatically add `extends: "eslint:recommended"` for easy quick start

### PR DESCRIPTION
**Pull request by gcochard**
~~This is a WIP so I would like some feedback before actually merging.~~

~~I have made a few changes regarding the docs, that I think will be superseded by #2915. I think I'd be more comfortable rebasing on top of #2915 when it lands. I have also updated the tests to comply with this behavior but I'm guessing #2915 will do that better as well.~~

Rebased on #2915, ready to go.

### CODE REVIEW

File: conf/environments.js [hunk 1]
```diff
@@ -13,20 +13,7 @@ module.exports = {
         globals: globals.browser
     },
     node: {
-        globals: globals.node,
-        ecmaFeatures: {
```
Comment (ilyavolodin) [inline conf/environments.js]: We still need globalReturn

File: conf/eslint.json [hunk 2]
```diff
@@ -10,11 +10,11 @@
     },
 
     "rules": {
-        "no-alert": 2,
-        "no-array-constructor": 2,
+        "no-alert": 0,
```
Comment (ilyavolodin) [inline conf/eslint.json]: Those should all still stay enabled. config.js should be modified to not use them unless "extends:eslint-recommended" is specified.

File: lib/cli.js [hunk 3]
```diff
@@ -30,6 +30,8 @@ var fs = require("fs"),
 
 debug = debug("eslint:cli");
 
+/*eslint-disable no-console*/
```
Comment (ilyavolodin) [inline lib/cli.js]: I think that's left over from testing?

File: lib/config-initializer.js [hunk 4]
```diff
@@ -44,6 +44,9 @@ function processAnswers(answers) {
     config.rules.quotes = [2, answers.quotes];
     config.rules["linebreak-style"] = [2, answers.linebreak];
     config.rules.semi = [2, answers.semi ? "always" : "never"];
+    if (answers.extendDefault) {
```
Comment (nzakas) [inline lib/config-initializer.js]: I'm wondering if we should just always add this without asking? Thoughts?

File: lib/config.js [hunk 5]
```diff
@@ -110,6 +110,12 @@ function loadConfig(filePath) {
             // Make the last element in an array take the highest precedence
             config = configExtends.reduceRight(function (previousValue, parentPath) {
 
+                if (parentPath === "eslint:recommended") {
+                    // Add an explicit substitution for eslint:recommended to conf/eslint.json
+                    // this lets us use the eslint.json file as the recommended rules
+                    parentPath = "./conf/eslint.json";
+                }
+
                 if (isFilePath(parentPath)) {
```
Comment (nzakas) [inline lib/config.js]: Should this be an else?

File: lib/options.js [hunk 6]
```diff
@@ -49,7 +49,7 @@ module.exports = optionator({
     }, {
         option: "reset",
         type: "Boolean",
-        default: "false",
+        default: "true",
```
Comment (nzakas) [inline lib/options.js]: This option will be removed, so no need to change here.

File: tests/lib/eslint.js [hunk 7]
```diff
@@ -2377,39 +2375,9 @@ describe("eslint", function() {
...
-                    globalReturn: false
-                }
-            }, filename);
-
-            assert.equal(messages.length, 1);
-            assert.equal(messages[0].message, "Illegal return statement");
-        });
-
-        it("should not parse global return when Node.js environment is false", function() {
+        it("should not parse global return when not passed ecmaFeatures", function() {
```
Comment (nzakas) [inline tests/lib/eslint.js]: Why this change?

File: lib/config.js (hunk 5 above)
Comment (gcochard) [inline lib/config.js]: Correct, good catch. I think this may be the source of a few unexpected test results I was seeing. Thanks for this.

File: lib/config-initializer.js (hunk 4 above)
Comment (gcochard) [inline lib/config-initializer.js]: Not sure about that. I think asking a bunch of style questions and then forcing other styles on the user might be a little unexpected. I think defaulting to true is probably the most we want to do here.

**Comment by gcochard** [comment]
@ilyavolodin Take another look, I'll try to get all the tests passing tonight, but the implementation is good.

File: lib/config-initializer.js (hunk 4 above)
Comment (ilyavolodin) [inline lib/config-initializer.js]: Agree with @gcochard I think it would be unexpected behavior for most users.

**Comment by ilyavolodin** [comment]
@gcochard Implementation looks good. Just fixes for unit tests left and I think it can be merged after that.

File: lib/config.js [hunk 8]
```diff
@@ -110,7 +110,11 @@ function loadConfig(filePath) {
             // Make the last element in an array take the highest precedence
             config = configExtends.reduceRight(function (previousValue, parentPath) {
 
-                if (isFilePath(parentPath)) {
+                if (parentPath === "eslint:recommended") {
+                    // Add an explicit substitution for eslint:recommended to conf/eslint.json
+                    // this lets us use the eslint.json file as the recommended rules
+                    parentPath = "./conf/eslint.json";
```
Comment (btmills) [inline lib/config.js]: Will this relative path work all the time? An example of what I'm thinking:

```sh
$ cd /Users/btmills/myproject
$ cat .eslintrc
{
    "extends": "eslint:recommended"
}
$ eslint test.js
```

Would `./conf/eslint.json` resolve to `/Users/btmills/myproject/conf/eslint.json` there?

File: lib/config-initializer.js (hunk 4 above)
Comment (nzakas) [inline lib/config-initializer.js]: The recommended rules will not be style rules, just potential errors (see issue description). Like, `no-unused-vars`, for instance.

File: lib/config.js (hunk 8 above)
Comment (nzakas) [inline lib/config.js]: Good catch! This would need to be resolved relative to __dirname.

File: lib/config-initializer.js (hunk 4 above)
Comment (gcochard) [inline lib/config-initializer.js]: Hmm, that's definitely a good point. I'm convinced default-on is fine. Anyone else have strong feelings either way on this?

**Comment by gcochard** [comment]
@ilyavolodin I'm having trouble with this; I have no idea why these 6 tests are failing. Any ideas?

**Comment by ilyavolodin** [comment]
Hmm.. I'm looking into them right now, and it's very strange, first one "should default defined variables to read-only" seems like an invalid test to me. It declares two global variables as read only, then test file overrides it and expects exit code to be 0 (even though there are linting errors). I'm not sure why that is. @btmills you added this test. I'm I not reading it correctly or something?

**Comment by ilyavolodin** [comment]
Tests under "when supplied with report output file path" are failing cause no rules are enabled, so no linting errors are returned and as such output file is not created. You should enable appropriate rules from the command line for those 3 tests.
"should return the project config when called in current working directory" is failing because the test is no longer valid. It's trying to compare content of the .eslintrc file from the root of ESLint to the same file, but processed by ConfigHelper (and since you added extends to root .eslintrc they are no longer the same, since actual will contain all recommended rules).
And "should merge command line config when config file adds environment to local .eslintrc" is failing for the opposite reason. You haven't added `extends` to override-env-conf.yaml

**Comment by gcochard** [comment]
@ilyavolodin @btmills I see the issue with the first test. I think I was *trying* to fix it, but I messed up the exit code. Changing the expected to `1` works there, and exhibits the expected behavior. Cli tests fixed now. On to the config tests!

**Comment by ilyavolodin** [comment]
@gcochard Ahh.. I didn't look at the changes you made, I assumed that the first failing test was not modified, so that would've made it invalid. But it makes much more sense that you modified it and just changed exit code:-)

**Comment by gcochard** [comment]
@ilyavolodin Have another look, if all is good I will squash and remove [wip] for a real PR!

File: tests/lib/eslint.js [hunk 9]
```diff
@@ -974,23 +974,6 @@ describe("eslint", function() {
             eslint.verify(code, {}, filename, true);
         });
 
-        it("should mark variables in Node.js environment as used", function() {
```
Comment (btmills) [inline tests/lib/eslint.js]: This test seems completely unrelated - why was it removed?

Comment (gcochard) [inline tests/lib/eslint.js]: I had commented it out in my first commit, and then mindlessly deleted it because it was commented out. I'll be adding it back in.

**Comment by ilyavolodin** [comment]
The changes looks good. One thing though, based on the issue that this is fixing (#2713) this doesn't address one of the bullet points. It doesn't change default eslint.json to disable all of the stylistic rules. Only actual code errors should be enabled as part of the `eslint:recommended`.
You can change (fixes #2713) to (refs #2713) and then create another pull request with updated eslint.json or you can just modify it as part of this PR, doesn't really matter.

**Comment by gcochard** [comment]
@ilyavolodin I disabled the stylistic rules, only the "Potential Errors" are in `conf/eslint.json` and by extension `eslint:recommended`. Ready to merge!

**Comment by ilyavolodin** [comment]
Awesome! Thanks for sticking with it @gcochard LGTM. @nzakas when you get a chance, can you verify list of recommended rules before I merge?

**Comment by btmills** [comment]
Since this disables stylistic rules, it should re-enable them in the project .eslintrc.

**Comment by gcochard** [comment]
Agreed, the rules that were enabled by default should carry over, correct?

**Comment by btmills** [comment]
Yeah, no reason to make changes, especially since @IanVS is taking care of the two deprecated ones in #2964.

**Comment by gcochard** [comment]
Per @btmills' suggestion, I re-enabled the rules in the project `.eslintrc`. All tests passing again.

**Comment by btmills** [comment]
LGTM

**Comment by nzakas** [comment]
@gcochard can you rebase onto master?

**Comment by gcochard** [comment]
Yep, I'll do it later tonight.

File: conf/eslint.json [hunk 10]
```diff
@@ -1,18 +1,18 @@
...
+        "no-caller": 0,
+        "no-catch-shadow": 0,
         "no-cond-assign": 2,
         "no-console": 2,
         "no-constant-condition": 2,
         "no-continue": 0,
         "no-control-regex": 2,
         "no-debugger": 2,
-        "no-delete-var": 2,
+        "no-delete-var": 0,
```
Comment (nzakas) [inline conf/eslint.json]: I think we should keep this on

File: conf/eslint.json [hunk 11]
```diff
@@ -21,86 +21,86 @@
...
-        "no-extra-bind": 2,
+        "no-extend-native": 0,
+        "no-extra-bind": 0,
         "no-extra-boolean-cast": 2,
         "no-extra-parens": 0,
         "no-extra-semi": 2,
-        "no-extra-strict": 2,
-        "no-fallthrough": 2,
+        "no-extra-strict": 0,
+        "no-fallthrough": 0,
```
Comment (nzakas) [inline conf/eslint.json]: And this

File: conf/eslint.json [hunk 12]
```diff
@@ -21,86 +21,86 @@
...
+        "no-iterator": 0,
+        "no-label-var": 0,
+        "no-labels": 0,
+        "no-lone-blocks": 0,
         "no-lonely-if": 0,
-        "no-loop-func": 2,
+        "no-loop-func": 0,
         "no-mixed-requires": [0, false],
-        "no-mixed-spaces-and-tabs": [2, false],
+        "no-mixed-spaces-and-tabs": [0, false],
```
Comment (nzakas) [inline conf/eslint.json]: And this

File: conf/eslint.json [hunk 13]
```diff
@@ -21,86 +21,86 @@
...
+        "no-new": 0,
+        "no-new-func": 0,
+        "no-new-object": 0,
         "no-new-require": 0,
-        "no-new-wrappers": 2,
+        "no-new-wrappers": 0,
         "no-obj-calls": 2,
-        "no-octal": 2,
-        "no-octal-escape": 2,
+        "no-octal": 0,
```
Comment (nzakas) [inline conf/eslint.json]: And this

File: conf/eslint.json [hunk 14]
```diff
@@ -21,86 +21,86 @@
...
         "no-param-reassign": 0,
         "no-path-concat": 0,
         "no-plusplus": 0,
         "no-process-env": 0,
-        "no-process-exit": 2,
-        "no-proto": 2,
-        "no-redeclare": 2,
+        "no-process-exit": 0,
+        "no-proto": 0,
+        "no-redeclare": 0,
```
Comment (nzakas) [inline conf/eslint.json]: And this

File: conf/eslint.json [hunk 15]
```diff
@@ -21,86 +21,86 @@
...
         "no-sparse-arrays": 2,
         "no-sync": 0,
         "no-ternary": 0,
-        "no-trailing-spaces": 2,
+        "no-trailing-spaces": 0,
         "no-this-before-super": 0,
         "no-throw-literal": 0,
-        "no-undef": 2,
-        "no-undef-init": 2,
+        "no-undef": 0,
```
Comment (nzakas) [inline conf/eslint.json]: And this.

File: conf/eslint.json [hunk 16]
```diff
@@ -21,86 +21,86 @@
...
         "no-unexpected-multiline": 0,
-        "no-underscore-dangle": 2,
+        "no-underscore-dangle": 0,
         "no-unneeded-ternary": 0,
         "no-unreachable": 2,
-        "no-unused-expressions": 2,
-        "no-unused-vars": [2, {"vars": "all", "args": "after-used"}],
-        "no-use-before-define": 2,
+        "no-unused-expressions": 0,
+        "no-unused-vars": [0, {"vars": "all", "args": "after-used"}],
```
Comment (nzakas) [inline conf/eslint.json]: And this

**Comment by gcochard** [comment]
@nzakas, should I move those to the "potential errors" section in the docs or no?

**Comment by nzakas** [comment]
I don't think it matters too much - the categorization is somewhat arbitrary.

**Comment by ilyavolodin** [comment]
@gcochard Can you squash commits and rebase please?

**Comment by gcochard** [comment]
@ilyavolodin Squashed, rebased, and ready to go again.

**Comment by ilyavolodin** [comment]
Awesome! Thanks a lot @gcochard for your patience!
